### PR TITLE
Emulation of RISC-V H extension

### PR DESCRIPTION
**Pull request by francois141**

### CODE REVIEW

**Comment by francois141** [comment]
After 5 hours of gdb.... green!!!! @CharlyCst You can review :-)

**Review by CharlyCst — COMMENTED** [review]
There are a few things to fix here and there, in particular in the switch from payload to firmware (see comments), but otherwise looks good to me :)

File: firmware/hypervisor/main.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            "csrw htval, t3",          // Write back to htval
+
+            // Hypervisor virtual interrupt enable
+            "csrr t4, hvip",           // Read Hypervisor Virtual Interrupt Pending Register
+            "csrw hvip, t4",           // Write back to hvip
+
+            // Insert fences for hypervisor virtual memory synchronization
+            // HFENCE.VVMA vaddr, asid
+            "li t5, 0",                // Load immediate 0 into t7 (representing ASID = 0 for all guests)
+            "hfence.vvma zero, t5",    // Synchronize virtual memory for all guest virtual machines
```
[the remark this answers is not in the document]
Comment (francois141) [reply firmware/hypervisor/main.rs:49]: You are right :-)

File: src/arch/metal.rs [hunk 2]
```diff
@@ -400,6 +446,56 @@ impl Architecture for MetalArch {
         }
     }
 
+    unsafe fn hfencegvma(vaddr: Option<usize>, asid: Option<usize>) {
```
[the remark this answers is not in the document]
Comment (francois141) [reply src/arch/metal.rs:449]: No, I never add any issue with them

File: src/decoder.rs [hunk 3]
```diff
@@ -376,7 +397,7 @@ fn decode_csr(csr: usize) -> Csr {
         0xF15 => Csr::Mconfigptr,
         0x302 => {
             if !Plat::HAS_S_MODE {
-                log::info!(
+                log::error!(
```
[the remark this answers is not in the document]
Comment (francois141) [reply src/decoder.rs:400]: Let's switch to warn!

**Review by CharlyCst — APPROVED** [review]
Awesome looks all good to me now :)

I'll rebase and merge the PR

**Pull request merged**